File: src/yacht.js
const score = (dice, category) => {
  const groupMap = dice.reduce((map, die) => {
    const newTotal = map.get(die) + die || die;
    map.set(die, newTotal);
    return map;
  }, new Map());

  const groups = groupMap.size;
  const sorted = [...dice].sort();
  const entries = [...groupMap];
  const anyLoneDice = entries.some(([k, v]) => k === v);

  const calculateFullHouse = () => groups === 2 && !anyLoneDice && sum(dice);

  const calculateFourOfAKind = () =>
    (groups === 1 && sum(dice.slice(1))) ||
    (groups === 2 && anyLoneDice && entries.find(([k, v]) => k !== v)[1]);

  const deepEq = (arr1, arr2) => arr1.every((el, i) => el === arr2[i]);
  const sum = arr => arr.reduce((a, c) => a + c, 0);

  return (
    {
      ones: groupMap.get(1),
      twos: groupMap.get(2),
      threes: groupMap.get(3),
      fours: groupMap.get(4),
      fives: groupMap.get(5),
      sixes: groupMap.get(6),
      yacht: groups === 1 && 50,
      'full house': calculateFullHouse(),
      'four of a kind': calculateFourOfAKind(),
      'little straight': deepEq(sorted, [1, 2, 3, 4, 5]) && 30,
      'big straight': deepEq(sorted, [2, 3, 4, 5, 6]) && 30,
      choice: sum(dice),
    }[category] || 0
  );
};

module.exports = { score };

File: go/exercism/yacht/yacht.go
package yacht

import (
	"sort"
)

func makeMap(dice []int) map[int]int {
	uniq := make(map[int]int)

	for _, num := range dice {
		// uniq[num] is initialized to 0
		uniq[num] += 1
	}

	return uniq
}

func sum(dice []int) int {
	total := 0
	for _, num := range dice {
		total += num
	}
	return total
}

func testYacht(dice []int) int {
	initial := dice[0]
	for _, num := range dice {
		if num != initial {
			return 0
		}
	}

	return 50
}

func testSingles(value int, dice []int) int {
	total := 0
	for _, num := range dice {
		if num == value {
			total += num
		}
	}
	return total
}

func testFull(dice []int) int {
	uniq := makeMap(dice)
	total := sum(dice)

	for _, v := range uniq {
		if !(v == 2 || v == 3) {
			return 0
		}
	}
	return total
}

func testFourOfAKind(dice []int) int {
	uniq := makeMap(dice)

	// TODO: Can this be cleaner?
	for k, v := range uniq {
		if !(v == 1 || v == 4 || v == 5) {
			return 0
		}
		if v == 4 || v == 5 {
			return k * 4
		}
	}
	return 0
}

func testStraight(size string, dice []int) int {
	prev := dice[0]

	if (size == "little" && dice[0] == 2) || (size == "big" && dice[0] == 1) {
		return 0
	}

	for _, num := range dice[1:] {
		if num != prev+1 {
			return 0
		}
		prev = num
	}

	return 30
}

func Score(dice []int, category string) int {
	sort.Ints(dice)
	// fmt.Println("category", category)
	switch category {
	case "yacht":
		return testYacht(dice)
	case "ones":
		return testSingles(1, dice)
	case "twos":
		return testSingles(2, dice)
	case "threes":
		return testSingles(3, dice)
	case "fours":
		return testSingles(4, dice)
	case "fives":
		return testSingles(5, dice)
	case "sixes":
		return testSingles(6, dice)
	case "full house":
		return testFull(dice)
	case "four of a kind":
		return testFourOfAKind(dice)
	case "little straight":
		return testStraight("little", dice)
	case "big straight":
		return testStraight("big", dice)
	case "choice":
		return sum(dice)
	default:
		return 0
	}
}
